drivers/can/socket_can_generic.h

```diff
@@ -118,38 +197,52 @@ static struct canbus_api socket_can_api = {
 
 static struct socket_can_context socket_can_context_1;
 
+static void state_changed(enum can_state state, struct can_bus_err_cnt err_cnt)
```

Comment:
Nut sure what you are doing here. Why do you want to have an Rx and Tx state? The state is the state of the controller. It doesn't mater which error counter is bigger.

Reply:
I've created the this software according to the socketcan standard, which is implemented in the linux kernel.
According to the errors on the bus, the controller enters a warning (which is not implemented) or a passive mode.

> A node starts out in Error Active mode. When any one of the two Error Counters raises above 127, the node will enter a state known as Error Passive and when the Transmit Error Counter raises above 255, the node will enter the Bus Off state. (Source: https://www.kvaser.com/about-can/the-can-protocol/can-error-handling/)  

This is not quite complete:
>  The state is the state of the controller

It is the state according to the counters, so the counters are part of the error report. And provides the application the necessary information on the bus. A high bus load could lead to `CAN_ERR_CRTL_RX_PASSIVE`. But if your overloading your hardware transceiver , or it's (electrically) broken you might receive `CAN_ERR_CRTL_TX_PASSIVE`.

You might want to check the documentation and existing source code.

Reply:
The error counters do only increase on bit errors. This is according to the Bosch CAN specification [1] chapter six. The states you are describing are exactly the same as defined in the specification.
You can just translate the state to the flags used in SocketCAN

[1] http://esd.cs.ucr.edu/webres/can20.pdf

Reply:
> When a RECEIVER detects an error, the RECEIVE ERROR COUNT will be
increased by 1, except when the detected error was a BIT ERROR during the
sending of an ACTIVE ERROR FLAG or an OVERLOAD FLAG.

This doesn't say it's only increased on bit errors.

Reply:
@maxvankessel Sorry, I was a little bit unclear on this. The counters only increase on errors detected, defined in chapter six. But anyway, the state you get from `enum can_state state` is the estate you should pass on.

Reply:
It's still unclear to me what you want. 

I think the `can_state state` is passed to the `socket_can_change_state` and I do think the controller tells us there are (at least) three error possibilities `CAN_ERR_CRTL_RX_PASSIVE`, `CAN_ERR_CRTL_TX_PASSIVE` & `CAN_ERR_BUSOFF`. A generic way to find this problem is to compare error counters. Counters are not passed on.

This seems to be the way how it's done in the linux kernel.

But please, provide a suggestion.

Reply:
Now I got it. There are two different kinds of states defined in the Linux kernel, depending on which counter exceeded 128.